src/hotspot/share/gc/g1/g1FullGCCompactionPoint.cpp

```diff
@@ -102,6 +103,7 @@ void G1FullGCCompactionPoint::forward(oop object, size_t size) {
 
   // Store a forwarding pointer if the object should be moved.
   if (cast_from_oop<HeapWord*>(object) != _compaction_top) {
+    preserved_stack()->push_if_necessary(object, object->mark());
```

Comment:
Can this be made conditionally on whether the markword is NOT marked/forwarded? (IOW, move the added predicate in `markWord` here.) The rationale is to minimize changes to the shared code for G1 specific usages.

Reply:
My previous msg was probably unclear... I meant sth like:

```c
if (!object->is_forwarded) {
  preserved_stack()->push_if_necessary...
}
```

(The newly added assert here is incorrect -- fails in GHA.)

Reply:
Oh I see! This makes more sense, yes. I made that change and cleaned up non G1 code.

BTW: I am working on new/alternative full-GCs for Serial, G1 and Shenandoah that don't store forwarding pointers in the mark-word at all, and thus avoid all that preserved-headers stuff (and together with the OM-world work that some Oracle engineers are working on, get rid of header displacement altogether). The SerialGC version is almost ready in Lilliput: https://github.com/openjdk/lilliput/pull/122